### PR TITLE
Base patch changes for tower/ruins randomisation

### PR DESCRIPTION
**Pull request by Darrman**
This adds locations for each floor of the Tower of Valni and Lagdou Ruins to FE8 Archipelago. This PR is in conjunction with https://github.com/CT075/Archipelago/pull/6 on the APWorld repo.

### CODE REVIEW

File: bin/connector_config/Generate.hs [hunk 1]
```diff
@@ -42,33 +42,42 @@ data Chapter
     | -- For now, we treat Eirika and Ephraim versions of the same map as the
       -- same location
       C Int
+    | T Int
+    | R Int
     | C5x
     | Endgame
     | Victory
+    | R10
```
Comment (CT075) [inline bin/connector_config/Generate.hs]: This doesn't need to be its own variant, we can just use the existing `R`.

File: bin/connector_config/Generate.hs [hunk 2]
```diff
@@ -42,33 +42,42 @@ data Chapter
...
     toEnum 6 = C5x
     toEnum 22 = Endgame
     toEnum 23 = Victory
+    toEnum 41 = R10
     toEnum i
         | i < 1 = error $ "invalid chapter index " ++ show i
         | i < 6 = C i
         | i < 22 = C (i - 1)
+        | i < 32 = T (i)
+        | i < 41 = R (i)
```
Comment (CT075) [inline bin/connector_config/Generate.hs]: This is wrong, we want index 24 to map to Tower 1, but as written this maps to Tower 24 (which doesn't exist).

File: bin/connector_config/Generate.hs [hunk 3]
```diff
@@ -217,6 +226,11 @@ emitConnectorConfigH emitLn = do
     emitLn $ "#define PrologueId (" ++ show (fromEnum $ Prologue) ++ ")"
     emitLn $ "#define EndgameId (" ++ show (fromEnum $ Endgame) ++ ")"
     emitLn $ "#define VictoryId (" ++ show (fromEnum $ Victory) ++ ")"
+    forM_ [24 .. 31] $ \i ->
+        emitLn $ "#define Tower" ++ show (i - 23) ++ "Id (" ++ show (fromEnum $ T i) ++ ")"
+    forM_ [32 .. 40] $ \i ->
+        emitLn $ "#define Ruins" ++ show (i - 31) ++ "Id (" ++ show (fromEnum $ R i) ++ ")"
```
Comment (CT075) [inline bin/connector_config/Generate.hs]: We should be doing the `i-31` in `toEnum` and `fromEnum` instead of here.

**Review by CT075 — CHANGES_REQUESTED** [review]

File: bin/connector_config/Generate.hs (hunk 3 above)
Comment (Darrman) [reply bin/connector_config/Generate.hs]: The lines are adjusted a bit to fix range errors, but the ``i-31`` is still around here as even though it's done in ``to`` and ``fromEnum`` leaving it as ``i`` just results in Tower 24 anyway.

**Pull request merged**